packages/instruction-plans/src/transaction-planner-base.ts

```diff
@@ -0,0 +1,416 @@
+    SOLANA_ERROR__INSTRUCTION_PLANS__CANNOT_PACK_USING_PROVIDED_MESSAGE,
+    SOLANA_ERROR__INSTRUCTION_PLANS__FAILED_TO_FIT_INSTRUCTION_PLAN_IN_NEW_MESSAGE,
+    SOLANA_ERROR__INSTRUCTION_PLANS__INVALID_INSTRUCTION_PLAN_KIND,
+    SOLANA_ERROR__INSTRUCTION_PLANS__INVALID_TRANSACTION_PLAN_KIND,
+    SOLANA_ERROR__INSTRUCTION_PLANS__MISSING_INSTRUCTIONS_IN_INSTRUCTION_PLAN,
+    SolanaError,
+} from '@solana/errors';
+import {
+    appendTransactionMessageInstructions,
+    BaseTransactionMessage,
+    TransactionMessageWithFeePayer,
+} from '@solana/transaction-messages';
+import { getTransactionMessageSize, TRANSACTION_SIZE_LIMIT } from '@solana/transactions';
+
+import {
+    InstructionPlan,
+    MessagePackerInstructionPlan,
+    ParallelInstructionPlan,
+    SequentialInstructionPlan,
+    SingleInstructionPlan,
+} from './instruction-plan';
+import {
+    getAllSingleTransactionPlans,
+    nonDivisibleSequentialTransactionPlan,
+    parallelTransactionPlan,
+    sequentialTransactionPlan,
+    SingleTransactionPlan,
+    singleTransactionPlan,
+    TransactionPlan,
+} from './transaction-plan';
+import type { TransactionPlanner } from './transaction-planner';
+
+type Mutable<T> = { -readonly [P in keyof T]: T[P] };
+
+type CreateTransactionMessage = (config?: {
+    abortSignal?: AbortSignal;
+}) =>
+    | Promise<BaseTransactionMessage & TransactionMessageWithFeePayer>
+    | (BaseTransactionMessage & TransactionMessageWithFeePayer);
+
+type OnTransactionMessageUpdated = <
+    TTransactionMessage extends BaseTransactionMessage & TransactionMessageWithFeePayer,
+>(
+    transactionMessage: TTransactionMessage,
+    config?: { abortSignal?: AbortSignal },
+) => Promise<TTransactionMessage> | TTransactionMessage;
+
+/**
+ * Configuration object for creating a new transaction planner.
+ *
+ * The `createTransactionMessage` function is used to create new transaction messages
+ * whenver needed and the `onTransactionMessageUpdated` function is an optional callback
+ * that can be used to update transaction messages during the planning process.
+ *
+ * @see {@link createBaseTransactionPlanner}
+ */
+export type TransactionPlannerConfig = {
+    /** Called whenever a new transaction message is needed. */
+    createTransactionMessage: CreateTransactionMessage;
+    /** Called whenever a transaction message is updated — e.g. new instructions were added. */
+    onTransactionMessageUpdated?: OnTransactionMessageUpdated;
+};
+
+/**
+ * Creates a new transaction planner based on the provided configuration.
+ *
+ * At the very least, the `createTransactionMessage` function must be provided.
+ * This function is used to create new transaction messages whenever needed.
+ *
+ * Additionally, the `onTransactionMessageUpdated` function can be provided
+ * to update transaction messages during the planning process. This function will
+ * be called whenever a transaction message is updated, e.g. when new instructions
+ * are added to a transaction message.
```

Comment:
I think it's important to drive home that the `onTransactionMessageUpdated` receives the candidate transaction message as an argument, but in all cases _must_ return a transaction message, modified or not.